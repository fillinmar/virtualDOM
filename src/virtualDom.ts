import { setAttributes, updateAttributes } from './attributes';

type Props = {
    [key: string]: any;
};
export type VNode =
    | { type: string; props: Props; children: VNode[] }
    | string;

export const createElement = (node: VNode): HTMLElement | Text => {
    if (typeof node === 'string') {
        return document.createTextNode(node);
    }
    const element = document.createElement(node.type as any);
    node.props && setAttributes(element, node.props);

    node.children.forEach(child => {
        const childElement = createElement(child);
        element.appendChild(childElement);
    });

    return element;
};

const hasChanged = (node1: VNode, node2: VNode) => {
    if (typeof node1 === 'string' || typeof node2 === 'string') {
        return node1 !== node2;
    } else {
        return node1.type !== node2.type;
    }
};

export const updateElement = (
    parent: HTMLElement,
    oldNode: VNode | null,
    newNode: VNode,
    index: number = 0
) => {
    if (!oldNode) {
        newNode && parent.appendChild(createElement(newNode));
    } else if (!newNode) {
        parent.removeChild(parent.childNodes[index]);
    } else if (hasChanged(oldNode, newNode)) {
        parent.replaceChild(createElement(newNode), parent.childNodes[index]);
    } else if (typeof newNode !== 'string' && typeof oldNode !== 'string') {
        updateAttributes(
            parent.childNodes[index] as HTMLElement,
            oldNode.props,
            newNode.props
        );

        const oldLength = oldNode.children.length;
        const newLength = newNode.children.length;

        for (let i = 0; i < newLength || i < oldLength; i++) {
            // We pass in parent.childNodes[index] instead of parent because we want to update only a specific child node of the parent element, rather than the entire parent element and all of its children.
            //     In virtual DOM, when an element needs to be updated, we create a new element with the updated properties and compare it with the old element to find the differences. We then update only the parts of the DOM that have changed.
            //     By passing in the specific child node at the given index, we ensure that only that node is updated, while the rest of the nodes in the parent element remain unchanged. If we passed in the entire parent element, the entire element would be updated, including its children, which could potentially be inefficient if only a small portion of the element needs to be updated.

            //findDOMNode  https://reactjs.org/docs/react-dom.html#render
            updateElement(
                parent.childNodes[index] as HTMLElement,
                oldNode.children[i],
                newNode.children[i],
                i
            );
        }
    }
};
