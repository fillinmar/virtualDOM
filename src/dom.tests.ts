import { createElement, updateElement, VNode} from './virtualDom';

describe('virtualDom', () => {
	describe('createElement', () => {
		test('Создаёт одну текстовую ноду', () => {
			expect(createElement('It works!')).toMatchSnapshot();
		});
		test('Создаёт один элемент', () => {
			expect(createElement({type: 'DIV', props: {}, children: []})).toMatchSnapshot();
		});

		test('Создаёт один элемент с двумя атрибутами', () => {
			expect(createElement({
				type: 'DIV',
				props: {
					style: 'background-color: red;',
					'tab-index': 0
				},
				children: [],
			})).toMatchSnapshot();
		});

		test('Создаёт один элемент с двумя детьми', () => {
			expect(createElement({
				type: 'DIV',
				props: {},
				children: [
					{type: 'SPAN', props: {}, children: []},
					'Some text'
				]
			})).toMatchSnapshot();
		});
	});


	describe('updateElement', () => {
		let container: HTMLDivElement;

		beforeEach(() => {
			container = document.createElement('div');
			document.body.appendChild(container);
		});

		afterEach(() => {
			container.remove();
		});

		it('updates a DOM node with a new VNode', () => {
			const oldVNode: VNode = {
				type: 'DIV',
				props: {className: 'old'},
				children: [],
			};
			const newVNode: VNode = {
				type: 'DIV',
				props: {className: 'new'},
				children: [],
			};

			const oldElement = createElement(oldVNode);

			container.appendChild(oldElement);

			updateElement(container, oldVNode, newVNode);
			expect(container.firstChild).toStrictEqual(createElement(newVNode));
		});

		it('updates a DOM node with new children', () => {
			const oldVNode: VNode = {
				type: 'div',
				props: {},
				children: ['foo'],
			};
			const newVNode: VNode = {
				type: 'div',
				props: {},
				children: ['bar'],
			};

			container.appendChild(createElement(oldVNode));

			updateElement(container, oldVNode, newVNode);

			expect(container.firstChild).toStrictEqual(createElement(newVNode));
			expect(container.firstChild?.textContent).toBe('bar');
		});

		it('adds new child elements', () => {
			const oldVNode: VNode = {
				type: 'div',
				props: {},
				children: ['first'],
			};
			const newVNode: VNode = {
				type: 'div',
				props: {},
				children: ['first', 'second'],
			};

			const oldElement = createElement(oldVNode);
			container.appendChild(oldElement);

			updateElement(container, oldVNode, newVNode);

			expect(container.firstChild).toBe(oldElement);
			expect(container.firstChild?.textContent).toBe('firstsecond');
		});

		it('removes child elements', () => {
			const oldVNode: VNode = {
				type: 'div',
				props: {},
				children: ['first', 'second'],
			};
			const newVNode: VNode = {
				type: 'div',
				props: {},
				children: ['first'],
			};

			const oldElement = createElement(oldVNode);
			container.appendChild(oldElement);

			updateElement(container, oldVNode, newVNode);

			expect(container.firstChild).toBe(oldElement);
			expect(container.firstChild?.textContent).toBe('first');
		});

		test('Обновляет детей 1 -> 1 разные типы', () => {
			const oldNode = {
				type: 'DIV',
				props: {},
				children: [
					'Some text',
				]
			}
			const oldElement = createElement(oldNode);
			// соответствует Node.TEXT_NODE
			expect(oldElement.firstChild?.nodeType).toBe(3);

			container.appendChild(oldElement);
			updateElement(container, oldNode, {
				type: 'DIV',
				props: {},
				children: [
					{type: 'P', props: {}, children: []},
				]
			});

			// соответствует Node.ELEMENT_NODE
			expect(oldElement.firstChild?.nodeType).toBe(1);
		});
	});
	describe('attributes', () => {
		let container: HTMLDivElement;

		beforeEach(() => {
			container = document.createElement('div');
			document.body.appendChild(container);
		});

		afterEach(() => {
			container.remove();
		});

		test('Обновляет атрибуты 1 -> 2', () => {
			const oldNode = {
				type: 'DIV',
				props: {},
				children: [],
			}
			const oldElement = createElement(oldNode);
			container.appendChild(oldElement);

			updateElement(container, oldNode, {
				type: 'DIV',
				props: {style: {backgroundColor: 'black'}},
				children: [],
			});

			expect((oldElement as HTMLElement).getAttribute('style')).toBe("background-color: black;");
		});

		// TODO: add more tests
	});
});
