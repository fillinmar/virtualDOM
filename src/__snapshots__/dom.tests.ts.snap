// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`virtualDom createElement Создаёт один элемент 1`] = `<div />`;

exports[`virtualDom createElement Создаёт один элемент с двумя атрибутами 1`] = `
<div
  tab-index="0"
/>
`;

exports[`virtualDom createElement Создаёт один элемент с двумя детьми 1`] = `
<div>
  <span />
  Some text
</div>
`;

exports[`virtualDom createElement Создаёт одну текстовую ноду 1`] = `It works!`;
